src/PrettyPrinter.ml

```diff
@@ -39,12 +39,25 @@ module Doc : DOC = struct
 
   let in_kwd = !^"in"
 
-  let rec of_type = function
-    | Type.Unit -> unit_type
-    | Type.Base idT -> !^idT
-    | Type.Prod (t1, t2) -> parens (of_type t1 ^^ cross ^^ of_type t2)
-    | Type.Arr (dom, cod) -> parens (of_type dom ^^^ arrow ^^^ of_type cod)
-    | Type.Box t -> box_type ^^ of_type t
+  let enclose = function true -> parens | false -> fun x -> x
+
+  let rec of_type =
```

Comment:
The standard approach here is to parameterize a pretty printing function with *precedence level*. It results in a bit simpler code.
It's probably easier to get the point across by showing a code snippet (`p` stands for precedence level below):
```
  let of_type =
    let rec walk p = function
      | Type.Unit -> unit_type
      | Type.Base idT -> !^idT
      | Type.Prod (t1, t2) -> parens_if (p > 1) (walk 1 t1 ^^^ cross ^^^ walk 2 t2)
      | Type.Arr (dom, cod) -> parens_if (p > 0) (walk 1 dom ^^^ arrow ^^^ walk 0 cod)
      | Type.Box t -> box_type ^^ walk 2 t
    in
    walk 0
```
`Box` has highest precedence, hence precedence `2`, then goes `Prod` with precedence `1` and then `Arr` with p = 0.
Notice that we supply different precedence levels to the components of `Prod` and `Arr` basing on their associativity.